Add substitute_node_with_dag to c api

mtreinish opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds a new function qk_dag_substitute_node_with_dag to the C api. This facilitates in place substitute from the new dag api. This is fairly mechanical addition to the api as
`DAGCircuit::substitute_node_with_dag()` is already rust native. So we only need to expose that inteface to C. The only difference between the C and existing internal rust interface is that the qubit, clbit, and var mapping arguments are not exposed to C. These options aren't commonly used, and if the user needs to map the ordering they can do that during replacement dag construction rather that at substitution time.

### Details and comments

Fixes #15190

mtreinish committed:
Add substitute_node_with_dag to c api

This commit adds a new function qk_dag_substitute_node_with_dag to the C
api. This facilitates in place substitute from the new dag api. This is
fairly mechanical addition to the api as
`DAGCircuit::substitute_node_with_dag()` is already rust native. So
we only need to expose that inteface to C. The only difference between
the C and existing internal rust interface is that the qubit, clbit, and
var mapping arguments are not exposed to C. These options aren't
commonly used, and if the user needs to map the ordering they can do
that during replacement dag construction rather that at substitution
time.

Fixes #15190

jakelishman commented on crates/cext/src/dag.rs:
```diff
@@ -968,3 +971,96 @@ pub unsafe extern "C" fn qk_dag_topological_op_nodes(dag: *const DAGCircuit, out
+/// @param node The node index of the operation to replace with the other `QkDag`. This
+///     must be the node index for an operation node in ``dag`` and the qargs and cargs
+///     count must match the number of qubits and clbits in `replacement`.
+/// @param replacement The other `QkDag` to replace `node` with. This dag must have
+///     the same number of qubits as the operation for ``node``. The node
+///     bit ordering will be ordering will be handled in order, so `qargs[0]` for
+///     `node` will be mapped to `qubits[0]` in `replacement`, `qargs[1]` to
+///     `qubits[0]`, etc. The same pattern applies to classical bits too.
```
I didn't think about it in the meeting, but the reason that the map needs to exist is for nodes that have "extra" wires, like control-flow or classical expressions - those don't appear in `qargs` or `cargs`, so they don't have a native order that the user can be determined, and so they can't just build `replacement` in order.

Also, if there _are_ any `Var`s in either circuit, then `substitute_node_with_dag` panics if you pass `None` in that argument (??).

If we merge this PR without it, we'll likely have to break the API later to add it back in.  That's not _necessarily_ a deal-breaker, but it is more than we thought of at the time.

mtreinish replied:
I can expose the mapping interfaces as an array of `uint32_t` with `NULL` for no mapping (we don't even need a length because it has to match the qargs length). That should work for Vars too (although for vars it's a bit weird to expose that and not have it in C yet) right?

mtreinish replied:
As discussed offline we'll look at adding a new variant function like `substitute_node_with_dag_reordering` that takes the mapping for the control flow use case when we add that to the C API. But we will keep this function as the simpler entrypoint for the common case.

jakelishman commented on crates/cext/src/dag.rs:
```diff
@@ -968,3 +971,96 @@ pub unsafe extern "C" fn qk_dag_topological_op_nodes(dag: *const DAGCircuit, out
+pub unsafe extern "C" fn qk_dag_substitute_node_with_dag(
+    dag: *mut DAGCircuit,
+    node: u32,
+    replacement: *const DAGCircuit,
```
We maybe don't want to meaningfully change things in this PR since it stems from Rust, but is "borrow" semantics definitely the right choice for `replacement`?  From C it makes things awkward, because we have to manually free each replacement DAG, whereas if it was "move" semantics, then I think most cases would be more natural, and we'd be able to inline `qk_dag_copy` calls for those that weren't.

Not a big deal, since the Rust code already has the same interface, but seeing it written out again here made me think.

mtreinish replied:
I can make it take ownership of the dag and free it as part of the function. It's easy to do if you think that's a better interface for this. I just did it this way because it was how the rust interface worked. But there isn't a reason we can't take ownership with this function.

jakelishman replied:
I'm uncertain.  This does feel like it wants to be a "move" to me, but given we never really had that option in Python, I don't really trust my intuition around it yet.

jakelishman commented on crates/cext/src/dag.rs:
```diff
@@ -968,3 +971,96 @@ pub unsafe extern "C" fn qk_dag_topological_op_nodes(dag: *const DAGCircuit, out
+    node: u32,
+    replacement: *const DAGCircuit,
+    error: *mut *mut c_char,
+) -> ExitCode {
```
I think the only failure modes in this function are:
- Python-space stuff
- `replacement` having the wrong number of edges / wires / etc
- the maps being specified incorrectly

I think bullets 2 and 3 are things that we typically consider "user error" in C, and don't generally warrant a failure condition, and 1 shouldn't matter.

There's no huge harm here, since we're not otherwise _using_ the return value, but is it an interface we want to commit to?

mtreinish replied:
There is a third error mode that `node` is not in the dag or not an op node. I can make it a panic with a check up front for all of these if you'd prefer.

jakelishman replied:
`node` not being a valid node is already a panic in the majority of functions we have (c.f. `qk_circuit_get_instruction`, and all the DAG getters) - that one's under "user error" too.

I don't have a strong feeling yet for exactly where the line of error vs abort _should_ be in C - I guess it comes down to whether we want to penalise performance for the highest performance extension modules in order to allow extensions to build more "interactive" tooling manually (like Python).  Certainly I wasn't suggesting adding _new_ checks so much as just turning the "write out the error message" form into an "abort" form.

I could imagine in the future we introduce a separate "checked" API for many things, where the caller is responsible for checking the output of most methods (much more akin to Python), and have the "bare" functions be more like "if you call me wrong, I abort".  In cases where the input is sufficiently complex that the user can't realistically be expected to know it's valid (e.g. "will this circuit successfully compile to that backend") then error messages are immediately appropriate, but for most things, I feel like the bare API should be "abort", and we would add _new_ C API explicitly to support the kind of "called on untrusted user input and checked on every call" interactive use that Python represents.

mtreinish replied:
I've switched this to a panic in https://github.com/Qiskit/qiskit/pull/15374/commits/a09d8b4e201437dba4ffe0754ac8b74cb938a993

jakelishman commented on crates/cext/src/dag.rs:
```diff
@@ -968,3 +971,96 @@ pub unsafe extern "C" fn qk_dag_topological_op_nodes(dag: *const DAGCircuit, out
+                        err.backtrace()
+                    ))
+                    .unwrap()
+                    .into_raw();
```
1. this is clearly copy/pasted from the transpiler
2. can we move the creation of the `CString` and `*mut char` out of the `unsafe` block, since no part of that bit is `unsafe`

mtreinish replied:
1. Yes it is, since it felt like the best way to handle a `PyErr`
2. Sure I can move it outside the unsafe block and leave writing into the pointer the only thing in the block. That being said I think if we move to panics for user input error we can drop this code and the exit code return.

jakelishman replied:
The point of 1 was more like both the comment and the message are incorrect lol

mtreinish replied:
After https://github.com/Qiskit/qiskit/pull/15374/commits/a09d8b4e201437dba4ffe0754ac8b74cb938a993 this isn't relevant anymore.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into substituting_nodes_with_dags_at_sea

mtreinish committed:
Merge remote-tracking branch 'origin/main' into substituting_nodes_with_dags_at_sea

mtreinish committed:
Panic instead of returning an exit code on failure modes

The failure modes for calling the inner rust method are all caused by
invalid input or aren't possible via the C API (come from python) so
there isn't a reason to have a subpar error experience dealing with
backtraces when it's not something recoverable.

jakelishman committed:
Run clang format

jakelishman committed:
Update documentation

jakelishman reviewed (approved):
I'm happy just leaving be the discussions on move/borrow semantics of the incoming replacement DAGs, since we already don't have it in an ideal situation in Rust space - we can potentially revisit in the future.

I also note that the Rust-space method returns a mapping of node indices in the replacement DAG to the corresponding entries in the base one, but I think that if we choose to expose that to C, it can happily be in a different method.  Its existence provides API pressure to accept DAGs by reference (since you'd _probably_ want the old DAG around after to keep the node indices valid), which feels like something else to think about in Rust space too.

I fixed the lint and docs in a minor commit. Release note is deferred pending #15221.
